internal/aws/ecssd/docker_label.go

```diff
@@ -0,0 +1,111 @@
+// Copyright  OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ecssd
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ecs"
+	"go.uber.org/zap"
+)
+
+// DockerLabelConfig matches all tasks based on their docker label.
+//
+// NOTE: it's possible to make DockerLabelConfig part of CommonExporterConfig
+// and use it both ServiceConfig and TaskDefinitionConfig.
+// However, based on existing users, few people mix different types of filters.
+// If that usecase arises in the future, we can rewrite the top level docker lable filter
+// using a task definition filter with arn_pattern:*.
+type DockerLabelConfig struct {
+	CommonExporterConfig `mapstructure:",squash" yaml:",inline"`
+
+	// PortLabel is mandetory, empty string means docker label based match is skipped.
+	PortLabel        string `mapstructure:"port_label" yaml:"port_label"`
+	JobNameLabel     string `mapstructure:"job_name_label" yaml:"job_name_label"`
+	MetricsPathLabel string `mapstructure:"metrics_path_label" yaml:"metrics_path_label"`
+}
+
+func (d *DockerLabelConfig) Init() error {
+	// It's possible to support it in the future, but for now just fail at config,
+	// so user don't need to wonder which port is used in the exported target.
+	if len(d.MetricsPorts) != 0 {
```

Comment:
I don't want to split the shared config because docker label is just one of three matchers, though its behaviour is pretty different from service and task based matcher so we could give it another struct like `ExportConfigWithoutPorts`. I plan to delay the change for next pr when service & task are added.

Reply:
Actually I think since we are going to have https://github.com/open-telemetry/opentelemetry-collector/pull/2802 it's better to put the validate logic here instead of removing it from struct. Because if user use `metrics_ports` for docker label and it's not in the struct, it will throw unmarshal error without telling why it is not there while it is supported for other filters. Though in the long term I think docker label should become part of common export config as well, you can discover based on service name and pick port based on docker label in task definition, we are just trying to be consistent with cloudwatch agent's behaviour here at this moment.